ee/modules/021-cni-cilium/hooks/ee/types.go

```diff
@@ -0,0 +1,70 @@
+/*
+Copyright 2024 Flant JSC
+Licensed under the Deckhouse Platform Enterprise Edition (EE) license. See https://github.com/deckhouse/deckhouse/blob/main/ee/LICENSE
+*/
+
+package ee
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"sort"
+
+	"k8s.io/apimachinery/pkg/types"
+)
+
+var hashFunc func(key string) [32]byte
+
+func init() {
+	hashFunc = newHashFunc()
+}
+
+type EgressGatewayState struct {
+	Generation                  int64             `json:"generation"`
+	Name                        string            `json:"name"`
+	Mode                        string            `json:"mode"`
+	IP                          string            `json:"ip"`
+	RoutingTableName            string            `json:"routingTableName"`
+	UID                         types.UID         `json:"uid,omitempty"`
+	DesiredActiveNode           string            `json:"desiredActiveNode"`
+	AllNodes                    []string          `json:"allNodes"`
+	ReadyNodes                  []string          `json:"readyNodes"`
+	HealthyNodes                []string          `json:"healthyNodes"`
+	HealthyNodesWithEgressAgent []string          `json:"healthyNodesWithEgressAgent"`
+	CurrentActiveNodes          []string          `json:"currentActiveNodes"`
+	NodeSelector                map[string]string `json:"nodeSelector"`
+}
+
+func (eg *EgressGatewayState) electDesiredActiveNode() string {
+	if len(eg.ReadyNodes) == 0 {
+		return ""
+	}
+
+	// Algorithm: If the readyNodes slice is empty, then there is no sense to calculate desiredMaster.
+	// For each of readyNode, calculate the SHA256 hash of (EgressGatewayGroup.name + Node.Name).
+	// Add the hashes to an array and sort it. Select the first element from the resulting slice.
+
+	readyNodesCopy := make([]string, len(eg.ReadyNodes))
+	copy(readyNodesCopy, eg.ReadyNodes)
+
+	sort.Slice(readyNodesCopy, func(i, j int) bool {
+		hi := hashFunc(eg.Name + "#" + readyNodesCopy[i])
+		hj := hashFunc(eg.Name + "#" + readyNodesCopy[j])
+		return bytes.Compare(hi[:], hj[:]) < 0
```

Comment:
```suggestion
		return strings.Compare(readyNodesCopy[i], readyNodesCopy[j]) < 0
```

I didn't get the idea of sha256 here. 
1. Comparing raw strings does the same without hash calculation.
2. For sorting nodes for a single egress gateway group, the name is not required.

Reply:
Without hashing we won't achieve the traffic balancing between nodes.

i.e. There are three EgressGateways (eg0, eg1, eg2) and three ready Nodes (node0, node1, node2). Without hashing we will just sort the nodes and get the node0 as active for all the egs. With hashing we have a chance to spread the load between all the nodes.

Reply:
We stole the approach [from metallb](https://github.com/metallb/metallb/blob/83b5dd4739e32d22348524fbf7ffedc4f1a7ad1f/speaker/layer2_controller.go#L117).

Reply:
> i.e. There are three EgressGateways (eg0, eg1, eg2) and three ready Nodes (node0, node1, node2). Without hashing we will just sort the nodes and get the node0 as active for all the egs. With hashing we have a chance to spread the load between all the nodes.

It does make sense! Could you please leave a comment in the code for future generations of engineers? 🙂

Reply:
We left short comment describing the algorithm above 🙂